cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java

```diff
@@ -221,75 +278,72 @@ private void processMetrics() {
         MessageFetcher fetcher = messagingService.prepareFetch(topicIdMetaKey.getTopicId());
         fetcher.setLimit(fetcherLimit);
         if (lastMessageId != null) {
+          LOG.debug("Start fetching from lastMessageId = {}", lastMessageId);
           fetcher.setStartMessage(lastMessageId, false);
         } else {
+          LOG.debug("Start fetching from beginning");
           fetcher.setStartTime(0L);
         }
 
-        byte[] nextMessageId = null;
+        byte[] currentMessageId = null;
         try (CloseableIterator<RawMessage> iterator = fetcher.fetch()) {
           while (iterator.hasNext() && isRunning()) {
             RawMessage input = iterator.next();
             try {
               payloadInput.reset(input.getPayload());
               MetricValues metricValues = recordReader.read(decoder, recordSchema);
               records.add(metricValues);
-              nextMessageId = input.getId();
+              currentMessageId = input.getId();
               LOG.trace("Received message {} with metrics: {}", Bytes.toStringBinary(lastMessageId), metricValues);
             } catch (IOException e) {
               LOG.warn("Failed to decode message to MetricValue. Skipped. {}", e.getMessage());
             }
           }
         }
 
-        if (records.isEmpty()) {
-          LOG.trace("No metrics record to process.");
+        if (currentMessageId != null) {
+          topicMessageIds.put(topicIdMetaKey, currentMessageId);
+          lastMessageId = currentMessageId;
+        }
+
+        // Skip persisting if the records is empty and the lastMessageId has already been persisted in the meta table
+        if (records.isEmpty() && lastMessageId != null
+          && Arrays.equals(lastPersistedMessageIdMap.get(topicIdMetaKey), lastMessageId)) {
```

Comment:
It stores what messageId's were persisted last time. Since persisting messageId of a topic can happen in another thread, so `lastPersistedMessageIdMap` is shared across different threads.